### PR TITLE
Adds new combat cybernetic implant

### PR DESCRIPTION
**Pull request by ghost**
## About The Pull Request

Adds arm empower implant to the game - a cybernetic implant that will make your punches deal 13 damage(normal human attacks deal between 1 and 10 damage), and throw your target away(doesn't work if you are lying down) like a baseball bat does. It is inserted into your arm, and it works only for the arm it is inserted into.

As an EMP effect it would cause you to fall under your target, or get paralyzed due to muscle spasms. 

## Why It's Good For The Game

Expands the implant system and adds more things to stuff into yourself if you want to roleplay as a cool robo man.

## Changelog

:cl:
add: Added a new cybernetic implant that increases punch damage of a human
/:cl:

### CODE REVIEW

**Comment by Striders13** [comment]
make it punch the user repeatedly when emp'd

**Comment by moocowswag** [comment]
STOP HITTING YOURSELF STOP HITTING YOURSELF

Me ion gunning a traitor with north star and punch implant.

**Comment by newfren** [comment]
doesn't increasing the damage a punch does do funny things to knockdown chances

**Comment by SuperSlayer0** [comment]
> doesn't increasing the damage a punch does do funny things to knockdown chances

It doesn't directly increase your attack damage, it replaces your weak hand attack with an attack that deals 13 damage and throws your target

**Comment by SuperSlayer0** [comment]
> make it punch the user repeatedly when emp'd

Sorry, but I have another idea for an EMP effect

**Comment by jughu** [comment]
> > doesn't increasing the damage a punch does do funny things to knockdown chances
> 
> It doesn't directly increase your attack damage, it replaces your weak hand attack with an attack that deals 13 damage and throws your target, so you wouldn't knockdown your target every hit

so you just punch them into a wall?

**Comment by SuperSlayer0** [comment]
> so you just punch them into a wall?

The throw on the attack doesn't stunlock people if you throw them into a wall, if you are asking about this.

**Comment by Wallemations** [comment]
Why do you have to specifically code a left & right version?

**Comment by SuperSlayer0** [comment]
> Why do you have to specifically code a left & right version?

As I understand, I can't make the implant have more then one possible zone/slot.

**Comment by necromanceranne** [comment]
> > Why do you have to specifically code a left & right version?
> 
> As I understand, I can't make the implant have more then one possible zone/slot.

This is also how toolsets work, FYI

**Comment by SuperSlayer0** [comment]
> > > Why do you have to specifically code a left & right version?
> > 
> > 
> > As I understand, I can't make the implant have more then one possible zone/slot.
> 
> This is also how toolsets work, FYI. for that, you just screwdriver them to swap which arm they can be placed into. So maybe steal some code from there.

Okay

**Comment by ArthuroPep** [comment]
> > > Why do you have to specifically code a left & right version?
> > 
> > 
> > As I understand, I can't make the implant have more then one possible zone/slot.
> 
> This is also how toolsets work, FYI. for that, you just screwdriver them to swap which arm they can be placed into. So maybe steal some code from there.

TIL

**Comment by Elli-Skala** [comment]
If a golem got this implant couldn't it actually lower damage done

**Comment by SuperSlayer0** [comment]
> If a golem got this implant couldn't it actually lower damage done

It would. This implant is designed for humans, not golems.

**Comment by DaydreamIQ** [comment]
Does this stack with hulk/Scarp or override their damage? And is there potential for an emag effect if not.

**Comment by SuperSlayer0** [comment]
> Does this stack with hulk/Scarp or override their damage? 

It overrides other damage, so it doesn't stack with hulk or martial arts.

**Comment by Helg2** [comment]
`/datum/design/cyberimp_muscle`
     `     name = "Left Arm Empower Implant"`
I think you forgot to change it.

**Comment by Helg2** [comment]
> It overrides other damage, so it doesn't stack with hulk or martial arts.
Won't it be also better if it doesnt override but just give minor buff? Like 4-6 damage instead of 13. Because it can be used for some gimmics and then it'll just worsen the situation.

**Review by san7890 — CHANGES_REQUESTED** [review]
This just doesn't seem very balanced, at least hulk dehumanizes you and gives you chunky fingers. (and I think that's still just not good)... this is just way too stacked in favor of the roboticist that lathes+implants this.

Either lock this to only being orderable through cargo, or perhaps some type of discrete autosurgeon traitory thing?

**Comment by SuperSlayer0** [comment]
> Either lock this to only being orderable through cargo, or perhaps some type of discrete autosurgeon traitory thing?

Okay, will change it to be locked to being orderable through cargo

File: code/modules/surgery/organs/augments_arms.dm [hunk 1]
```diff
@@ -334,3 +334,90 @@
...
+/obj/item/organ/internal/cyberimp/arm/muscle/proc/reboot()
+	organ_flags &= ~ORGAN_FAILING
+
+/obj/item/organ/internal/cyberimp/arm/muscle/proc/on_attack_hand(mob/living/carbon/human/source, atom/target, proximity, modifiers)
+	SIGNAL_HANDLER
+
+	if(source.get_active_hand() != source.get_bodypart(check_zone(zone)))
+		return
+	if(!proximity)
+		return
```
Comment (san7890) [inline code/modules/surgery/organs/augments_arms.dm:381]:
```suggestion
	if(source.get_active_hand() != source.get_bodypart(check_zone(zone)) || !proximity)
		return
```

Short-circuiting works better.

File: code/modules/surgery/organs/augments_arms.dm [hunk 2]
```diff
@@ -334,3 +334,90 @@
...
+	if(!source.combat_mode || LAZYACCESS(modifiers, RIGHT_CLICK))
+		return
+	if(isliving(target))
```
Comment (san7890) [inline code/modules/surgery/organs/augments_arms.dm:384]:
```suggestion
	if(!isliving(target))
		return //include whatever value you need to here
```

early return is much cleaner indentation wise.

File: code/modules/surgery/organs/augments_arms.dm [hunk 3]
```diff
@@ -334,3 +334,90 @@
...
+
+/obj/item/organ/internal/cyberimp/arm/muscle/emp_act(severity)
+	. = ..()
+	if((organ_flags & ORGAN_FAILING) || . & EMP_PROTECT_SELF)
+		return
+	organ_flags |= ORGAN_FAILING
+	addtimer(CALLBACK(src, PROC_REF(reboot)), 90 / severity)
+
+/obj/item/organ/internal/cyberimp/arm/muscle/proc/reboot()
+	organ_flags &= ~ORGAN_FAILING
```
Comment (san7890) [inline code/modules/surgery/organs/augments_arms.dm:373]: balloon alert like "your arm stops spasming!", perhaps?

File: code/modules/surgery/organs/augments_arms.dm [hunk 4]
```diff
@@ -334,3 +334,90 @@
...
+		playsound(living_target.loc, 'sound/weapons/punch1.ogg', 25, TRUE, -1)
+
+		living_target.apply_damage(punch_damage, BRUTE)
+
+		if(source.body_position != LYING_DOWN) //Throw them if we are standing
+			var/atom/throw_target = get_edge_target_turf(living_target, source.dir)
+			living_target.throw_at(throw_target, attack_throw_range, rand(throw_power_min,throw_power_max), source, gentle)
+
+		living_target.visible_message(span_danger("[source] [picked_hit_type]ed [living_target]!"), \
+					span_userdanger("You're [picked_hit_type]ed by [source]!"), span_hear("You hear a sickening sound of flesh hitting flesh!"), COMBAT_MESSAGE_RANGE, source)
```
Comment (san7890) [inline code/modules/surgery/organs/augments_arms.dm:418]:
```suggestion
		living_target.visible_message(span_danger("[source] [picked_hit_type]ed [living_target]!"), \
					span_userdanger("You're [picked_hit_type]ed by [source]!"), span_hear("You hear a sickening sound of flesh hitting flesh!"), COMBAT_MESSAGE_RANGE, source)
```
```suggestion
		living_target.visible_message(
			span_danger("[source] [picked_hit_type]ed [living_target]!"), 
			span_userdanger("You're [picked_hit_type]ed by [source]!"),
			span_hear("You hear a sickening sound of flesh hitting flesh!"),
			COMBAT_MESSAGE_RANGE,
			source,
		)
```
Preferred way to multiline. Also, name your last two arguments please.

File: code/modules/surgery/organs/augments_arms.dm [hunk 5]
```diff
@@ -334,3 +334,90 @@
...
+/obj/item/organ/internal/cyberimp/arm/muscle/Insert(mob/living/carbon/reciever, special = FALSE, drop_if_replaced = TRUE)
+	. = ..()
+	if(ishuman(reciever)) //Sorry, only humans
+		RegisterSignal(reciever, COMSIG_HUMAN_EARLY_UNARMED_ATTACK, PROC_REF(on_attack_hand))
+
+/obj/item/organ/internal/cyberimp/arm/muscle/Remove(mob/living/carbon/implant_owner, special = 0)
+	. = ..()
+	UnregisterSignal(implant_owner, COMSIG_HUMAN_EARLY_UNARMED_ATTACK)
+
+/obj/item/organ/internal/cyberimp/arm/muscle/emp_act(severity)
```
Comment (san7890) [inline code/modules/surgery/organs/augments_arms.dm:365]: should have a balloon_alert in here, something like:

```txt
balloon_alert(user, "your arm spasms wildly!")
```

just to be clear to the viewer that it is indeed affected

**Review by san7890 — CHANGES_REQUESTED** [review]

File: code/modules/surgery/organs/augments_arms.dm [hunk 6]
```diff
@@ -334,3 +334,90 @@
...
+
+	actions_types = list()
+
+	var/punch_damage = 13 ///The amount of damage dealt by your punches. Not really high, but better then normall punches.
+	var/gentle = TRUE //Will your throws be gentle or no
+	var/attack_throw_range = 1 ///How far away your attack will throw your oponent
+	var/throw_power_min = 1
+	var/throw_power_max = 4
+
+	var/emp_base_duration = 9 SECONDS ///How long will the implant malfunction if it is EMP'd
```
[the remark this answers is not in the document]
Comment (SuperSlayer0) [reply code/modules/surgery/organs/augments_arms.dm]: Done

File: code/modules/surgery/organs/augments_arms.dm (hunk 5 above)
Comment (SuperSlayer0) [reply code/modules/surgery/organs/augments_arms.dm:370]: Added it

File: code/modules/surgery/organs/augments_arms.dm (hunk 3 above)
Comment (SuperSlayer0) [reply code/modules/surgery/organs/augments_arms.dm:379]: Done

File: code/modules/surgery/organs/augments_arms.dm [hunk 7]
```diff
@@ -334,3 +334,101 @@
...
+		living_target.throw_at(throw_target, attack_throw_range, rand(throw_power_min,throw_power_max), source, gentle)
+
+	living_target.visible_message(
+		span_danger("[source] [picked_hit_type]ed [living_target]!"), 
+		span_userdanger("You're [picked_hit_type]ed by [source]!"),
+		span_hear("You hear a sickening sound of flesh hitting flesh!"),
+		COMBAT_MESSAGE_RANGE,
+		source,
+		)
+	to_chat(source, span_danger("You [picked_hit_type] [target]!"))
```
Comment (san7890) [inline code/modules/surgery/organs/augments_arms.dm:430]:
```suggestion
	)

	to_chat(source, span_danger("You [picked_hit_type] [target]!"))
```

formatting nit

File: code/modules/surgery/organs/augments_arms.dm [hunk 8]
```diff
@@ -334,3 +334,101 @@
...
+	if(!isliving(target))
+		return //include whatever value you need to here
+	var/mob/living/living_target = target
+
+	source.changeNext_move(CLICK_CD_MELEE)
+	var/picked_hit_type = pick("punch", "smash", "kick")
+
+	if(organ_flags & ORGAN_FAILING)
+		if(source.body_position != LYING_DOWN && living_target != source && prob(50))
+			to_chat(source, span_danger("You try to [picked_hit_type] [living_target], but loose your balance and fall!"))
```
Comment (san7890) [inline code/modules/surgery/organs/augments_arms.dm:398]:
```suggestion
			to_chat(source, span_danger("You try to [picked_hit_type] [living_target], but lose your balance and fall!"))
```

File: code/modules/surgery/organs/augments_arms.dm [hunk 9]
```diff
@@ -334,3 +334,101 @@
...
+
+/obj/item/organ/internal/cyberimp/arm/muscle/proc/on_attack_hand(mob/living/carbon/human/source, atom/target, proximity, modifiers)
+	SIGNAL_HANDLER
+
+	if(source.get_active_hand() != source.get_bodypart(check_zone(zone)) || !proximity)
+		return
+	if(!source.combat_mode || LAZYACCESS(modifiers, RIGHT_CLICK))
+		return
+	if(!isliving(target))
+		return //include whatever value you need to here
```
Comment (san7890) [inline code/modules/surgery/organs/augments_arms.dm:390]: I didn't mean include my comment here, I meant that if you need something like `COMPONENT_CANCEL_ATTACK_CHAIN` or something here that you should have it. At the very least, remove that comment. Sorry if that was confusing.

**Review by san7890 — CHANGES_REQUESTED** [review]
should be final nits?

File: code/modules/surgery/organs/augments_arms.dm (hunk 9 above)
Comment (SuperSlayer0) [reply code/modules/surgery/organs/augments_arms.dm:390]: This should just make the user do a normal attack_hand rather then canceling it

**Comment by Kubisopplay** [comment]
It allows you to stunlock and murder someone who was standing next to a wall, and has almost 0 downsides. Even being locked to order is very strong, should be antag only thing.

**Comment by nukularpower** [comment]
This is way too strong, it makes hulk look like a joke.   For one thing, damage is too high, and for the other, the stunlock is ridiculous in a time when stun combat is basically being phased out.   I cant understand why this was merged, frankly - the crew does not need this sort of powerboost versus the weak TG antags.

**Comment by optimumtact** [comment]
@san7890 fix this you ape

**Comment by SuperSlayer0** [comment]
> It allows you to stunlock and murder someone who was standing next to a wall, and has almost 0 downsides. Even being locked to order is very strong, should be antag only thing.

Stunlock is an unintended thing, i will probably try to fix it

**Comment by Kubisopplay** [comment]
I kinda don't understand why crew has this sort of an implant at all, its straight up combat thing that you can order roundstart. If it gave something like being much faster at dragging or something it would be justified tool in crews hands, but right now its basically a weaker free powerfist that you cant drop and dont need to charge.

**Comment by nukularpower** [comment]
>As for damage, you can litterally get a better weapon just by using a protolathe(saws, cleavers with 15 damage and more wounding potencial)

That is exactly my point - this should not be doing MORE damage than a fricken cleaver

**Comment by jughu** [comment]
> > so you just punch them into a wall?
> 
> The throw on the attack doesn't stunlock people if you throw them into a wall, if you are asking about this.

🤨

**Comment by SuperSlayer0** [comment]
> That is exactly my point - this should not be doing MORE damage than a fricken cleaver

And it doesn't do more damage then a cleaver(13 vs 15 with 20 wound bonus)

> Also just tested, and this does stack with hulk. 3-4 punches to crit a monkey, 5-6 to kill it, nice!

Already PRed a fix of this

**Pull request merged**